js/capture.js

```diff
@@ -19,13 +27,61 @@ const capture = {
       });
   },
 
+  // In the parsed whitelist JSON object (from disconnect-entitylist.json):
+  //   - 'properties' keys are first parties
+  //   - 'resources' keys are third parties
+  reformatList(whiteList) {
+    const firstPartyWhiteList = {};
+    const thirdPartyWhiteList = {};
+    let counter = 0;
+    for (const siteOwner in whiteList) {
+      const firstParties = whiteList[siteOwner].properties;
+      for (let i = 0; i < firstParties.length; i++) {
+        firstPartyWhiteList[firstParties[i]] = counter;
+      }
+      const thirdParties = whiteList[siteOwner].resources;
+      thirdPartyWhiteList[counter] = [];
+      for (let i = 0; i < thirdParties.length; i++) {
+        thirdPartyWhiteList[counter].push(thirdParties[i]);
+      }
+      counter++;
+    }
+
+    return {
+      firstPartyWhiteList,
+      thirdPartyWhiteList
+    };
+  },
+
   // Returns true if the request should be stored, otherwise false.
-  shouldStore(tab) {
+  shouldStore(tab, thirdPartyFromRequest) {
     const documentUrl = new URL(tab.url);
     // ignore about:*, moz-extension:* & non-visible tabs (like dev tools)
     if (documentUrl.protocol !== 'about:'
       && documentUrl.protocol !== 'moz-extension:'
       && tab.id !== browser.tabs.TAB_ID_NONE) {
+      // check if third party is on the whitelist (owned by the first party)
+      if (thirdPartyFromRequest) {
+        const firstPartyFromRequest = documentUrl.hostname;
+        for (const firstParty in this.firstPartyWhiteList) {
+          const firstPartyFromList = new RegExp(firstParty);
```

Comment:
Per @jonathanKingston (paraphrasing)

Looks like the [safe browsing documentation](https://developers.google.com/safe-browsing/v4/urls-hashing) (the a.b.c example) suggests a basic solution to how to look at the lists:
* take developers.google.com and check every . into the name is in the list. Such that: developers.google.co.uk would check:
  - developers.google.co.uk
  - google.co.uk
  - co.uk

You pass through the list as many times as you have '.'s minus 1.
it's still (1) wrong for right to left and (2) overly expensive for companies that have extensions with .'s in them

What I mean by (1):
* hostnames are reversed in some languages
* assume this is in chinese: uk.co.thing.subdomain
* technically I think some hostnames can have mixed left to right and right to left which can screw up code too
* and there is also extensions that are right to left and hostnames that are left to right

What I mean by (2):
* as in .co.uk has an extra lookup
* some have 3 components too like .com.co.uk

That way you can use object/array lookups and only modify the URL coming in.